Kernel/Devices/AHCIPort.cpp

```diff
@@ -0,0 +1,508 @@
+#define SATA_SIG_PM 0x96690101
+
+#define HBA_PORT_IPM_ACTIVE 1
+#define HBA_PORT_DET_PRESENT 3
+
+#define ATA_DEV_BUSY 0x80
+#define ATA_DEV_DRQ 0x08
+#define ATA_CMD_READ_DMA_EX 0x25
+#define ATA_CMD_IDENTIFY 0xEC
+
+#define HBA_PxCMD_ST 0x0001
+#define HBA_PxCMD_FRE 0x0010
+#define HBA_PxCMD_FR 0x4000
+#define HBA_PxCMD_CR 0x8000
+#define HBA_PxIS_TFES 1 << 30
+
+#define SLOT_CFL 0x1F
+#define SLOT_ATAPI 0x1 << 5
+#define SLOT_WRITE 0x1 << 6
+#define SLOT_PREFETCH 0x1 << 7
+
+#define SLOT_RESET 0x1 << 8
+#define SLOT_BIST 0x1 << 9
+#define SLOT_CLEAR 0x1 << 10
+#define SLOT_PORT_MULT 0xF000
+#define DATA_ENTRY_DBC 0x3fffff
+#define DATA_ENTRY_I 0x1 << 31
+#define FIS_H2D_PMPORT 0xF
+#define FIS_H2D_MODE 0x1 << 7
+
+#define LSB(x) ((x)&0xFF)
+#define MSB(x) (((x) >> 8) & 0xFF)
+
+// Only store at least one page inside the buffer
+#define SECTOR_SIZE 512
+#define MAX_SECTORS (PAGE_SIZE / SECTOR_SIZE) + 1
+#define BUFFER_SIZE MAX_SECTORS* SECTOR_SIZE
+
+#define COMMAND_TABLE_SIZE 256
+#define COMMAND_LIST_SIZE 1024                    // 1K
+#define FIS_SIZE 1024 * 4                         // 4K
+#define COMMAND_SLOT_SIZE COMMAND_TABLE_SIZE * 32 // 8K
+
+namespace Kernel {
+
+struct FISRegHostToDevice {
+    u8 fis_type;
+
+    u8 byte1;
+    u8 command;
+    u8 featurel;
+
+    u8 lba0;
+    u8 lba1;
+    u8 lba2;
+    u8 device;
+
+    u8 lba3;
+    u8 lba4;
+    u8 lba5;
+    u8 featureh;
+
+    u8 countl;
+    u8 counth;
+    u8 icc;
+    u8 control;
+
+    u8 rsv1[4];
+};
+
+AHCISlot::AHCISlot(VirtualAddress base, AHCIPort& port, VirtualAddress buffer, PhysicalAddress buffer_physical, u32 index)
+    : m_base_addr(base)
+    , m_buffer(buffer)
+    , m_buffer_physical(buffer_physical)
+    , m_index(index)
+{
+    m_reg = (SlotReg*)m_base_addr.get();
+    rebase(port);
+}
+
+void AHCISlot::rebase(AHCIPort& port)
+{
+    m_reg->prdtl = 8;
+
+    // Command Table
+    // Offset:      5K + (256 * slot)
+    // Size:		256
+    // Max Count:	1
+    // Max Size:	256 * 1 (256)
+    m_ctba_offset = (COMMAND_LIST_SIZE + FIS_SIZE) + (COMMAND_TABLE_SIZE * m_index);
+    m_reg->ctba = port.base_physical() + m_ctba_offset;
+    m_reg->ctbau = 0;
+
+    auto table = (u8*)(port.base() + m_ctba_offset);
+    m_cfis = table + 0x00;
+    m_acmd = table + 0x40;
+    m_entries = (DataEntry*)(table + 0x80);
+    memset(table, 0, COMMAND_TABLE_SIZE);
+}
+
+void AHCISlot::setup_command(AHCIPort& port)
+{
+    // Reset command table
+    memset((void*)(port.base() + m_ctba_offset), 0,
+        256);
+}
+
+bool AHCISlot::execute_command(AHCIPort& port)
+{
+    // Wait until port is not busy
+    u32 timeout = 0;
+    while (port.is_busy() && timeout < 1000000)
+        timeout += 1;
+
+    if (timeout == 1000000) {
+        dbg() << "AHCI: Port " << port.index() << " timed out\n";
+        ASSERT_NOT_REACHED();
+        return false;
+    }
+
+    // Issue the command
+    port.issue_command(m_index);
+
+    // Wait for command to be done
+    do {
```

Comment:
From what I see - this is a simple wait loop that doesn't rely on interrupts at all. Better to put the thread in a wait queue and let it wake when an interrupt arrives.